File: m/rls/filoc.r
file	filoc - convert to local specification
include	rid:rider
include	rid:fidef
If !Pdp
include	rid:stdef
include	rid:mxdef
include	rid:ctdef
include	rid:lndef
include	rid:medef
End

  proc	fi_loc
	spc : * char			; file spec
	loc : * char			; result area (can be same as spc)
  is
If Pdp
	st_cop (spc, loc)
Else
  	nxt : * char			; next character of interest
	cnt : int = 0			; count \'s
	quo : int = 0			; needs quotes wrapped at end

     repeat
	fi_trn (spc, loc, 0)		; translate before localisation

;	st_cop (spc, loc)		; return in result area

;	Remove versions (;xxx)

	nxt = st_fnd (";", loc)		; look for version
	*nxt = 0 if ne			; scrap versions

;	Convert (`) to space

	while (nxt=st_fnd("`", loc)) ne	; find pseudo space
	.. ++quo, *nxt = ' '		; replace with a space

;	xxx	Done if no colon (:)
;	x:	Done if single character device name

	nxt = st_fnd (":", loc)		; find colon
	quit if null			; no device
	quit if loc[1] eq ':'		; single char device name
					
;	Done if "con", "conin$" or "conout$"

	*nxt = 0			; terminate it a second
	if st_sam (loc, "con")		;
	|| st_sam (loc, "conin$")	;
	|| st_sam (loc, "conout$")	;
	.. quit				; leave it untouched
					;
;	Strings that begin with underline (_) have the
;	underline removed and have no further processing.

	if *loc eq '_'			; starts with "_"
	   me_mov(loc+1,loc,st_len(loc)); delete the underline
	.. quit				; and done

;	Otherwise (xx) is converted to (\xx\)

	*nxt = '\\'			; replace with '\'
	st_ins ("\\", loc)		; another up front
     never
	if quo
	   st_ins ("\"", loc)		; wrap in quotes
	.. st_ins ("\"", st_end (loc))	;
End
  end
If !Pdp
code	fi_trn -- translate file spec

  func	fi_trn
	spc : * char
	res : * char
	mod : int			; 1 =>
	()  : int			; fails if no space
  is	lft : [mxSPC] char		;
	rep : [mxSPC] char		;
	trm : * char			;
	idx : int			;
	cnt : int = 16			; maximum 16 levels of redirection
	fst : int = *spc		; save first character
	trn : int = 0			;
	st_cop (spc, res)		; default translation

      while cnt--			;
	st_cop (res, lft)		; get a copy
					;
	if (trm = st_fnd (":", lft)) ne	; find the colon
	   *trm++ = 0			; drop and skip colon
	else				;
	   trm = lft			; search for non-alphanumeric
	   while *trm			;
	.. .. fine if !ct_aln (*trm++)	; got punctuation
					;
	if !ln_trn (lft, rep, 0) 	; attempt translation
	   fine if trn			; already translated once
	   fine if !(mod & 1)		; nothing to translate
	.. reply -1			; didn't translate
	++trn				; remember we translated
					;
	st_len (rep) + st_len (trm)	;
	fail if that gt (mxSPC-4)	; no room for translation
					;
	if *trm				; filename following
	&& *st_lst (rep) ne '\\'	; no terminating '\'
	&& *trm ne '\\'			; and none coming
	.. st_app ("\\", rep)		; add one
					;
	st_app (trm, rep)		; append filename part
	st_cop (rep, res)		; bingo
	fine if !st_fnd (":", res)	; any more to translate?
;	fine if fst eq '@'		; these do not loop
     end
	fail
  end

End
end file

; to dos
;	d:f.t	=>	d:f.t		if #d = 1
;			\d\f.t		otherwise
;
;	[s]f.t		\s\f.t
;	[s.s]f.t	\s\s\f.t
;
;	con
;	conin$
;	conout$				no further translation
;	x`y		x y		"^" replaced by space
;	_x:		x:		no further translation
;	x:				single character devices accepted
;	xx:yyy		xx\yy
;	xx:		xx\
;
;	other		...\		
;
; to vms
;
;	d:f.t	=>	d:f.t
;	\s\f.t		[s]f.t
;	\s\s\f.t	[s.s]f.t
;
;	Program`Files
;
;
; fi_trn
;
;	Isolate device part and translate it
;
; ???	@...		no recursive translation
;
;	aaa		an alphanumeric string
;	xxx		any string
;	yyy		any string
;	
;	xxx:yyy		xxx is translated
;			'\' inserted as separator if necessary:
;			if xxx translates
;			if translation does not end with '\'
;			if yyy is non-null
;			if yyy does not start with '\'
;	aaa		aaa is translated
If vms
	if ferror (fil->Pfil) ne	; had an error
	.. im_rep ("E-File input error [%s]", fil->Aspc)
End

If vms
	if *mod eq 'w'			; a new file
	   fil->Pfil = fopen (spc, mod, "RFM=STM") ; create stream file
	else				; an old file
	.. fil->Pfil = fopen (spc, mod)	; attempt open
Else
	fil->Pfil = fi_opn (spc, mod,""); open it
End


File: m/rls/rlb/filoc.c
/* file -  filoc - convert to local specification */
#include "m:\rid\rider.h"
#include "m:\rid\fidef.h"
#include "m:\rid\stdef.h"
#include "m:\rid\ctdef.h"
#include "m:\rid\lndef.h"
#include "m:\rid\medef.h"
/* code -  fi_loc - convert spec for local rules */
void fi_loc(
char *spc ,
char *loc )
{ char *nxt ;
  int cnt = 0;
  fi_trn (spc, loc, 0);
#if Dos || Win || Wnt
  if ( (nxt = st_fnd (";", loc)) != 0) {*nxt = 0 ;}
  if( (nxt = st_fnd (":", loc)) == NULL)return;
  if( loc[1] == ':')return;
  *nxt = 0;
  if (( st_sam (loc, "con"))
  ||(st_sam (loc, "conin$"))
  ||(st_sam (loc, "conout$"))) {
    return; }
  if ( *loc == '_') {
    me_mov(loc+1,loc,st_len(loc));
    return; }
  *nxt = '\\';
  st_ins ("\\", loc);
#elif  0
  for(;;)  {
    if( (nxt = st_fnd ("\\", nxt)) == NULL)break;
    ++cnt;
  } 
  if( cnt == 0)return;
  nxt = st_fnd ('\\', loc);
  if( --cnt == 0){ *nxt = ':' ; return;}
  *nxt = '[';
  for(;;)  {
    nxt = st_fnd ("\\", nxt);
    if( --cnt == 0)return;
    *nxt = '.';
  } 
  *nxt = ']';
#endif 
} 
/* code -  fi_trn -- translate file spec */
int fi_trn(
char *spc ,
char *res ,
int mod )
{ char lft [128];
  char rep [128];
  char *trm ;
  int idx ;
  st_cop (spc, lft);
  st_cop (lft, res);
  if ( (trm = st_fnd (":", lft)) != 0) {
    *trm++ = 0;
    } else {
    trm = lft;
    while ( *trm) {
      if( !ct_aln (*trm++))return 1; } }
  if( (ln_trn (lft, rep, 0)) == 0)return 1;
  if( (st_len (rep) + st_len (trm)) > 124)return 0;
  if (( *trm)
  &&(*st_lst (rep) != '\\')
  &&(*trm != '\\')) {
    st_app ("\\", rep); }
  st_app (trm, rep);
  st_cop (rep, res);
  return 1;
} 
